File: uPython/a_ESP32_scripts/a_bencher.py
import gc

from utime import ticks_us, ticks_diff, ticks_cpu
import micropython, machine

Lang_name = "uPython"

class BenchTimer:
    def __init__(self):
        self.tStart = 0
        self.cStart = 0
        self.started = False
        self.tDuration = 0
        self.cDuration = 0
        
    def start(self):
        self.tStart = ticks_us()
        self.cStart = ticks_cpu()
        self.started = True
        
    def stop(self):
        self.tDuration = ticks_diff(ticks_us(), self.tStart)
        self.cDuration = ticks_cpu() - self.cStart 
        if self.started != True :
            print("Timer was not started duration is invalid")
        self.started = False
        
##-------------------------------------------------------------------------
class PrintType:
    Readable = 0
    Matlab = 1
    CSV = 2
    
def Print_header(ptype):
    if ptype == PrintType.Readable:
        print("|%10s|%10s|%10s|%10s|%10s|%10s|%10s|%10s" % ("Lang", "Test", "Freq(Mhz)", "Iter", "Length", "CPU cyc", "Time us", "Rez"))
    elif ptype == PrintType.CSV:
        print("%10s;%10s;%10s;%10s;%10s;%10s;%10s;%10s" % ("Lang", "Test", "Freq(Mhz)", "Iter", "Length", "CPU cyc", "Time us", "Rez"))
    elif ptype == PrintType.Matlab:
        print("PrintType MATLAB not implemented\r\n")
    else:
        print("No such printtype\r\n")

##-------------------------------------------------------------------------
class RezVerifcation:
    Match    = "OK"
    NoMatch  = "Err"
    NoStd    = "NoStd"
    
##-------------------------------------------------------------------------
    
class Tester:
    TestName    = ""
    TestLengths = []
    Niterations = 0
    PrintType   = PrintType.Readable
    RunFn       = []
    
    def Run(self):
        Freq = int(machine.freq()/1000000)
        timer = BenchTimer()
        for dataLen in self.TestLengths:
            for iters in range(self.Niterations):
                rez = self.RunFn(dataLen, timer)
                if self.PrintType == PrintType.Readable:
                    print("|%10s|%10s|%10s|%10s|%10s|%10s|%10s|%10s" % (Lang_name, self.TestName, Freq, iters+1, dataLen, timer.cDuration, timer.tDuration, rez))
                elif self.PrintType == PrintType.CSV:
                    print("%10s;%10s;%10s;%10s;%10s;%10s;%10s;%10s" % (Lang_name, self.TestName, Freq, iters+1, dataLen, timer.cDuration, timer.tDuration, rez))
                elif self.PrintType == PrintType.Matlab:
                    print("PrintType MATLAB not implemented\r\n")
                else:
                    print("No such printtype\r\n")
            
        

        
    

File: C-Cpp/src/a_bencher.cpp
#include "a_data.h"
#include "a_bencher.h"

void benchTimer_start(BenchTimer * self){
    self->tStart = esp_timer_get_time();
    self->cStart = xthal_get_ccount();
	self->started = true;
}

void benchTimer_stop(BenchTimer * self){
    self->tDuration = esp_timer_get_time() - self->tStart;
    self->cDuration = xthal_get_ccount() - self->cStart;
	if (self->started != true) {
		printf("Timer was not started duration is invalid");
	}
	self->started = false;
}

const char * printVerification(RezVerifcation ver){
    switch(ver){
        case RezVerifcation::Match: return "OK";
        case RezVerifcation::NoMatch: return "Err";
        case RezVerifcation::NoSTD: return "NoSTD";
        case RezVerifcation::NotRun: return "Not run";
        default: return "Bad Verificatio code";
    }
}

void bench_Print_header(PrintType ptype){
    switch(ptype){
        case PrintType::Readable: 
            printf("|%-10s|%-10s|%-10s|%-10s|%-10s|%-10s|%-10s|%-10s\r\n", "Lang", "Test", "Freq(Mhz)", "Iter", "Length", "CPU cyc", "Time us", "Rez");
            break;
        case PrintType::CSV:
            printf("%-10s;%-10s;%-10s;%-10s;%-10s;%-10s;%-10s;%-10s\r\n", "Lang", "Test", "Freq(Mhz)", "Iter", "Length", "CPU cyc", "Time us", "Rez");
            break;
        case PrintType::Matlab:
            printf("PrintType MATLAB not implemented\r\n");
            break;
    }
}

void bench_Run(Tester * self){
    uint32_t Freq = getCpuFrequencyMhz();
    BenchTimer timer;
    timer.cDuration = 0;
    timer.tDuration = 0;
    int i = 0; 
    while (self->TestLengths[i] != -1){
        int32_t dataLen = self->TestLengths[i++];
        int iters = 0;
        RezVerifcation rez = RezVerifcation::NotRun;
        unsigned intlevel = portENTER_CRITICAL_NESTED();
        rez = self->RunFn(dataLen, &timer);
        for (; iters < self->Niterations; iters++){
            rez = self->RunFn(dataLen, &timer);
            switch(self->Ptype) {
                case Readable:
                    printf("|%-10s|%-10s|%-10d|%-10d|%-10d|%-10d|%-10d|%-10s\r\n", Lang_name, self->TestName, Freq, iters+1, dataLen, timer.cDuration, timer.tDuration, printVerification(rez));
                    break;
                case CSV:
                    printf("%-10s;%-10s;%-10d;%-10d;%-10d;%-10d;%-10d;%-10s\r\n", Lang_name, self->TestName, Freq, iters+1, dataLen, timer.cDuration, timer.tDuration, printVerification(rez));
                    break;
                case Matlab:
                    printf("PrintType MATLAB not implemented\r\n");
                    break;
            }
        }
    }
}